### PR TITLE
feat(966): Banners support

### PR DESCRIPTION
**Pull request by jimgrund**
## Context

Add support to manage and display banners for announcements

## References

* Issue: https://github.com/screwdriver-cd/screwdriver/issues/966

### CODE REVIEW

File: lib/banner.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+const BaseModel = require('./base');
+
+class BannersModel extends BaseModel {
```
Comment (tkyi) [inline lib/banner.js]: BannerModel

File: lib/banner.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+     * @param {Number} config.dateCreated     Date the banner object was created
+     * @param {Number} config.isActive        Boolean to indicate whether banner should display
+     * @param {String} config.type            The type of banner (info|warn)
+     */
+    constructor(config) {
+        super('banner', config); // data-schema model name
+    }
+}
+
+module.exports = BannersModel;
```
Comment (tkyi) [inline lib/banner.js]: BannerModel

File: lib/bannerFactory.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     */
+    create(config) {
+        if (!config.type) {
+            config.type = 'info';
+        }
+        if (!config.isActive) {
+            config.isActive = false;
+        }
+
+        // config.dateCreated = (new Date()).toISOString();
```
Comment (tkyi) [inline lib/bannerFactory.js]: config.createTime*

File: lib/banner.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+const BaseModel = require('./base');
+
+class BannersModel extends BaseModel {
+    /**
+     * Construct a BannersModel object
+     * @method constructor
+     * @param {Object} config
+     * @param {String} config.message         The banner message
+     * @param {Number} config.createdBy       The ID of the user that created this banner
+     * @param {Number} config.dateCreated     Date the banner object was created
```
Comment (tkyi) [inline lib/banner.js]: config.createTime

**Comment by tkyi** [comment]
Looks good so far, just need to
- address minor comments
- update the `README.md` with Banners information
- add some tests for when `config.type` and `config.isActive` do not exist

File: README.md [hunk 5]
```diff
@@ -1232,6 +1232,87 @@ factory.get(id).then(model => {
...
+factory.create(config).then(model => {
+    // do stuff with banner model
+});
+```
+
+| Parameter          | Type    | Required | Description                                                          |
+| :------------------| :-------| :--------| :--------------------------------------------------------------------|
+| config             | Object  | Yes      | Configuration Object                                                 |
+| config.message     | String  | Yes      | Text of banner to be displayed                                       |
+| config.isActive    | Boolean | No       | Flag for whether banner should display (default: false)              |
```
Comment (tkyi) [inline README.md]: Could add another column to the table for defaults

File: README.md [hunk 6]
```diff
@@ -1232,6 +1232,87 @@ factory.get(id).then(model => {
...
+| config.type        | String  | No       | Type/Severity of banner message. Options: info,warn  (default: info) |
```
Comment (tkyi) [inline README.md]: Can you wrap `info` and `warn` in back ticks `

File: README.md [hunk 7]
```diff
@@ -1232,6 +1232,87 @@ factory.get(id).then(model => {
...
+
+#### Get
+Get a banner based on unique id of banner. 
+```js
+factory.get(id).then(model => {
+    // do stuff with collection model
```
Comment (tkyi) [inline README.md]: do stuff with `banner`  model*

File: README.md [hunk 8]
```diff
@@ -1232,6 +1232,87 @@ factory.get(id).then(model => {
...
+});
+```
+
+| Parameter | Type   | Description                      |
+| :---------| :------| :--------------------------------|
+| id        | Number | The unique id for the collection |
```
Comment (tkyi) [inline README.md]: banner*

File: lib/banner.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+const BaseModel = require('./base');
+
+class BannerModel extends BaseModel {
+    /**
+     * Construct a BannersModel object
```
Comment (tkyi) [inline lib/banner.js]: BannerModel*

**Review by tkyi — APPROVED** [review]
Looks good to me! 👍 🎉

File: lib/bannerFactory.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+     * Create a Banner model
+     * @param {Object}  config
+     * @param {String}  config.message             The banner message
+     * @param {String}  config.createdBy           The username of the associated user
+     * @param {String}  config.createTime          The time the banner is created
+     * @param {Boolean} [config.isActive=false]     Whether the banner is active
+     * @param {String}  [config.type='info']       Type of banner (info|warn|etc)
+     * @memberof BannerFactory
+     */
+    create(config) {
```
Comment (minzcmu) [inline lib/bannerFactory.js]: why we do not generate the createTime inside this function? just like this https://github.com/screwdriver-cd/models/blob/master/lib/pipelineFactory.js#L52

Comment (tkyi) [reply lib/bannerFactory.js]: Oh that's weird I saw it commented out earlier

File: README.md [hunk 11]
```diff
@@ -1232,6 +1232,87 @@ factory.get(id).then(model => {
...
+    // do stuff with banner model
+});
+```
+
+| Parameter          | Type    | Required | Default | Description                                               |
+| :------------------| :-------| :--------| :-------| :---------------------------------------------------------|
+| config             | Object  | Yes      |         | Configuration Object                                      |
+| config.message     | String  | Yes      |         | Text of banner to be displayed                            |
+| config.isActive    | Boolean | No       | false   | Flag for whether banner should display                    |
+| config.type        | String  | No       | info    | Type/Severity of banner message. Options: `info`,`warn`   |
```
Comment (d2lam) [inline README.md:1271]: Missing a couple arguments here

Comment (jimgrund) [reply README.md:1271]: Which arguments are you referring to?

Comment (d2lam) [reply README.md:1271]: `createdBy`

File: lib/bannerFactory.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    createClass(config) {
+        return new Banner(config);
+    }
+
+    /**
+     * Create a Banner model
+     * @param {Object}  config
+     * @param {String}  config.message             The banner message
+     * @param {String}  config.createdBy           The username of the associated user
+     * @param {String}  config.createTime          The time the banner is created
```
Comment (tkyi) [inline lib/bannerFactory.js]: You can just remove the `createTime` comment here

**Review by minzcmu — APPROVED** [review]

**Pull request merged**